feat(OCM-invites): Implementation of invitation flow for OCM 1.1.0

mickenordin opened the pull request:
* Resolves: #38886

## Summary

This patchset implements the /invite-accepted endpoint

https://cs3org.github.io/OCM-API/docs.html?branch=v1.1.0&repo=OCM-API&user=cs3org#/paths/~1invite-accepted/post

A companion PR with UI for creating and otherwise lifecycle manage OCM invites is in the works for the Contacts app, but that work needs this first.

nickvergessen commented on apps/cloud_federation_api/lib/Controller/RequestHandlerController.php:
```diff
@@ -213,6 +219,84 @@ public function addShare($shareWith, $name, $description, $providerId, $owner, $
+		if (!$found_for_this_user) {
+			$response = ['message' => 'Invalid or non existing token', 'error' => true];
+			$status = Http::STATUS_BAD_REQUEST;
+			return new JSONResponse($response,$status);
```
```suggestion
			$response = new JSONResponse($response, $status);
			$response->throttle();
			return $response;
```

nickvergessen commented on apps/cloud_federation_api/lib/Controller/RequestHandlerController.php:
```diff
@@ -213,6 +219,84 @@ public function addShare($shareWith, $name, $description, $providerId, $owner, $
+		$response = ['userID' => $data['user_id'], 'email' => $sharedFromEmail, 'name' => $sharedFromDisplayName];
+		$status = Http::STATUS_OK;
+		$updated = new DateTime("now");
+		$qb->update('federated_invites f')
```
```suggestion
		$qb->update('federated_invites', 'f')
```
But not sure why you use an alias at all?

nickvergessen commented on apps/cloud_federation_api/lib/Controller/RequestHandlerController.php:
```diff
@@ -213,6 +219,84 @@ public function addShare($shareWith, $name, $description, $providerId, $owner, $
+			->set('f.recipient_name', $qb->createNamedParameter($name))
+			->set('f.recipient_user_id', $qb->createNamedParameter($userId))
+			->set('f.recipient_provider', $qb->createNamedParameter($recipientProvider))
+			->where($qb->expr()->eq('token', $qb->createNamedParameter($token)));
```
when you stay with the alias, use it here as well:
```suggestion
			->where($qb->expr()->eq('f.token', $qb->createNamedParameter($token)));
```

mickenordin replied:
I removed the use of the alias

nickvergessen commented on apps/cloud_federation_api/lib/Controller/RequestHandlerController.php:
```diff
@@ -213,6 +219,84 @@ public function addShare($shareWith, $name, $description, $providerId, $owner, $
+			->set('f.recipient_user_id', $qb->createNamedParameter($userId))
+			->set('f.recipient_provider', $qb->createNamedParameter($recipientProvider))
+			->where($qb->expr()->eq('token', $qb->createNamedParameter($token)));
+		$result = $qb->executeQuery();
+		$result->closeCursor();
```
This is not a select, so you should run statement and don't use a cursor

```suggestion
		$qb->executeStatement();
```

nickvergessen commented on apps/cloud_federation_api/lib/Migration/Version1015Date202502262004.php:
```diff
@@ -0,0 +1,96 @@
+declare(strict_types=1);
+
+/**
+ * SPDX-FileCopyrightText: 2019 Nextcloud GmbH and Nextcloud contributors
```
```suggestion
 * SPDX-FileCopyrightText: 2025 Nextcloud GmbH and Nextcloud contributors
```

nickvergessen commented on lib/private/OCM/Model/OCMProvider.php:
```diff
@@ -31,7 +35,10 @@ class OCMProvider implements IOCMProvider {
+		IConfig $config,
 	) {
+		$this->config = $config;
+		$this->provider = 'Nextcloud ' . $config->getSystemValue('version');
```
Would make more sense to expose the API version this implements instead of the internal 4 digit Nextcloud version?

mickenordin replied:
The spec says:

provider: A friendly branding name of this endpoint

https://cs3org.github.io/OCM-API/docs.html?branch=v1.1.0&repo=OCM-API&user=cs3org#/paths/~1ocm-provider/get

So this was my gut feeling, but I am open to other suggestions. The version of the api is exposed as apiVersion, so I dont know?

nickvergessen replied:
Then we should skip the version

nickvergessen commented on apps/cloud_federation_api/lib/Migration/Version1015Date202502262004.php:
```diff
@@ -0,0 +1,96 @@
+			]);
+
+			$table->addColumn('user_id', Types::STRING, [
+				'notnull' => false,
```
Did you mix this up?
```suggestion
				'notnull' => true,
```

User ID can not be empty.

nickvergessen commented on apps/cloud_federation_api/lib/Migration/Version1015Date202502262004.php:
```diff
@@ -0,0 +1,96 @@
+			// https://saturncloud.io/blog/what-is-the-maximum-length-of-a-url-in-different-browsers/#maximum-url-length-in-different-browsers
+			// We use the least common denominator, the minimum length supported by browsers
+			$table->addColumn('recipient_provider', Types::STRING, [
+				'notnull' => true,
```
This means we can also not store empty strings, is that intended?

mickenordin replied:
No, you are right, they probably need to be empty until the invite has been accepted, after that they should be populated. I will fix this.

nickvergessen commented on apps/cloud_federation_api/lib/Migration/Version1015Date202502262004.php:
```diff
@@ -0,0 +1,96 @@
+				'length' => 2083,
+			]);
+			$table->addColumn('recipient_user_id', Types::STRING, [
+				'notnull' => true,
```
This means we can also not store empty strings, is that intended?

mickenordin replied:
No, you are right, they probably need to be empty until the invite has been accepted, after that they should be populated. I will fix this.

nickvergessen commented on apps/cloud_federation_api/lib/Migration/Version1015Date202502262004.php:
```diff
@@ -0,0 +1,96 @@
+				'length' => 1024,
+			]);
+			$table->addColumn('recipient_name', Types::STRING, [
+				'notnull' => true,
```
This means we can also not store empty strings, is that intended?

mickenordin replied:
No, you are right, they probably need to be empty until the invite has been accepted, after that they should be populated. I will fix this.

nickvergessen commented on apps/cloud_federation_api/lib/Migration/Version1015Date202502262004.php:
```diff
@@ -0,0 +1,96 @@
+			]);
+			// https://www.directedignorance.com/blog/maximum-length-of-email-address
+			$table->addColumn('email', Types::STRING, [
+				'notnull' => true,
```
This means we can also not store empty strings, is that intended?

mickenordin replied:
No, you are right, they probably need to be empty until the invite has been accepted, after that they should be populated. I will fix this.

nickvergessen reviewed (dismissed):
Where is this table written?
I only see a single SELECT and an UPDATE, but no INSERT?

mickenordin commented:
> Where is this table written? I only see a single SELECT and an UPDATE, but no INSERT?

The ui part of this, where you can create and delete invites are coming in a companion pr for the contacts app.

nickvergessen commented on apps/cloud_federation_api/lib/Migration/Version1015Date202502262004.php:
```diff
@@ -0,0 +1,96 @@
+use OCP\Migration\IOutput;
+use OCP\Migration\SimpleMigrationStep;
+
+class Version1015Date202502262004 extends SimpleMigrationStep
```
Sorry, missed it before. When adding a migration you have to bump the version in the apps `appinfo/info.xml`

mickenordin replied:
I think this is correct now?

mickenordin replied:
After rename I get:

```
{"reqId":"hzaJctCpyi52NiP6tSDZ","level":3,"time":"2025-03-04T08:29:34+00:00","remoteAddr":"10.88.0.9","user":"--","app":"index","method":"POST","url":"/","message":"Migration step 'OCA\\CloudFederationAPI\\Migration\\Version1016Date202502262004' is unknown","userAgent":"Mozilla/5.0 (X11; Linux x86_64; rv:135.0) Gecko/20100101 Firefox/135.0","version":"32.0.0.0","exception":{"Exception":"InvalidArgumentException","Message":"Migration step 'OCA\\CloudFederationAPI\\Migration\\Version1016Date202502262004' is unknown","Code":0,"Trace":[{"file":"/var/www/html/lib/private/DB/MigrationService.php","line":419,"function":"createInstance","class":"OC\\DB\\MigrationService","type":"->","args":["*** sensitive parameters replaced ***"]},{"file":"/var/www/html/lib/private/DB/MigrationService.php","line":387,"function":"migrateSchemaOnly","class":"OC\\DB\\MigrationService","type":"->","args":["latest"]},{"file":"/var/www/html/lib/private/Installer.php","line":551,"function":"migrate","class":"OC\\DB\\MigrationService","type":"->","args":["latest",true]},{"file":"/var/www/html/lib/private/Installer.php","line":517,"function":"installShippedApp","class":"OC\\Installer","type":"::","args":["cloud_federation_api",null]},{"file":"/var/www/html/lib/private/Setup.php","line":362,"function":"installShippedApps","class":"OC\\Installer","type":"::","args":[false,null]},{"file":"/var/www/html/core/Controller/SetupController.php","line":44,"function":"install","class":"OC\\Setup","type":"->","args":["*** sensitive parameters replaced ***"]},{"file":"/var/www/html/lib/base.php","line":982,"function":"run","class":"OC\\Core\\Controller\\SetupController","type":"->","args":["*** sensitive parameters replaced ***"]},{"file":"/var/www/html/index.php","line":24,"function":"handleRequest","class":"OC","type":"::","args":[]}],"File":"/var/www/html/lib/private/DB/MigrationService.php","Line":481,"message":"Migration step 'OCA\\CloudFederationAPI\\Migration\\Version1016Date202502262004' is unknown","exception":{},"CustomMessage":"Migration step 'OCA\\CloudFederationAPI\\Migration\\Version1016Date202502262004' is unknown"}}
```
It worked before?

mickenordin replied:
The problem must be that I didn't run:
```
build/autoloaderchecker.sh
```
after rename, fixing that now.

mickenordin commented:
When working on the front end part, I realized we need an event to react to, so I added that.

ChristophWurst commented:
Hi @mickenordin,
thanks a lot for this contributions.

To continue after https://github.com/nextcloud/server/pull/51113#discussion_r1979049010, **do** commit the changes in the class maps, so composer finds new/moved files. **Do not** commit the InstalledVersions.php and similar metadata files. These look slightly different depending on the composer version you have installed on your system.

mickenordin commented:
> Hi @mickenordin, thanks a lot for this contributions.
> 
> To continue after [#51113 (comment)](https://github.com/nextcloud/server/pull/51113#discussion_r1979049010), **do** commit the changes in the class maps, so composer finds new/moved files. **Do not** commit the InstalledVersions.php and similar metadata files. These look slightly different depending on the composer version you have installed on your system.

Sorry about that, should be fixed now.

come-nc commented on apps/cloud_federation_api/lib/Controller/RequestHandlerController.php:
```diff
@@ -431,7 +550,8 @@ private function confirmNotificationIdentity(
 	 * @return void
 	 * @throws IncomingRequestException
 	 */
-	private function confirmNotificationEntry(?IIncomingSignedRequest $signedRequest, string $entry): void {
+	private function confirmNotificationEntry(?IIncomingSignedRequest $signedRequest, string $entry): void
+	{
```
Why are all opening brackets moved to a new line? I think `composer run cs:fix` should put them back on their line, no?

come-nc commented on apps/cloud_federation_api/lib/Migration/Version1016Date202502262004.php:
```diff
@@ -0,0 +1,96 @@
+			$table->setPrimaryKey(['id']);
+		}
+
+		return $schema;
```
Please only return `$schema` if modifications hapenned, return `null` otherwise.

come-nc commented on lib/public/OCM/IOCMProvider.php:
```diff
@@ -54,6 +54,25 @@ public function setApiVersion(string $apiVersion): static;
+	 * returns the capabilities of the API
+	 *
+	 * @return array
+	 * @since 30.0.0
```
```suggestion
	 * @since 32.0.0
```

Same for the others

come-nc commented on apps/cloud_federation_api/lib/Controller/RequestHandlerController.php:
```diff
@@ -213,6 +228,105 @@ public function addShare($shareWith, $name, $description, $providerId, $owner, $
+
+		/** @var IQueryBuilder $qb */
+		$qb = $this->db->getQueryBuilder();
+		$qb->select('*')
+			->from('federated_invites')
+			->where($qb->expr()->eq('token', $qb->createNamedParameter($token)));
```
There should an entity and a mapper for this table, having db requests in the controller is discouraged.

mickenordin replied:
Thanks for your review. Do you have a good example for me to look at on how to implement that according to best practices?

miaulalala replied:
From the top of my head, you can take a look at the Mail app or Talk (Spreed) app for some examples:

https://github.com/nextcloud/mail/blob/main/lib/Db/AliasMapper.php
https://github.com/nextcloud/mail/blob/main/lib/Db/Alias.php

or

https://github.com/nextcloud/spreed/blob/main/lib/Model/ReminderMapper.php
https://github.com/nextcloud/spreed/blob/main/lib/Model/Reminder.php

come-nc commented on apps/cloud_federation_api/lib/Db/FederatedInviteMapper.php:
```diff
@@ -0,0 +1,96 @@
+		return $this->mapToInvite($data);
+	}
+
+	public function mapToInvite(array $data): FederatedInvite {
+		$federatedInvite = new FederatedInvite();
+		$federatedInvite->setAccepted($data['accepted']);
+		$federatedInvite->setAcceptedAt($data['acceptedAt']);
+		$federatedInvite->setCreatedAt($data['createdAt']);
+		$federatedInvite->setExpiredAt($data['expiredAt']);
+		$federatedInvite->setRecipientEmail($data['recipientEmail']);
+		$federatedInvite->setRecipientName($data['recipientName']);
+		$federatedInvite->setRecipientProvider($data['recipientProvider']);
+		$federatedInvite->setRecipientUserId($data['recipientUserId']);
+		$federatedInvite->setToken($data['token']);
+		$federatedInvite->setUserId($data['userId']);
+		return $federatedInvite;
+	}
```
Same here, you should use $this->mapRowToEntity, or ideally $this->findEntity instead of implementing your own method.

mickenordin replied:
done

come-nc reviewed (commented):
I think because you defined getters/setters you short circtuited the entity/mapper system and had to re-implement stuff.

Look at https://github.com/nextcloud/server/blob/master/apps/webhook_listeners/lib/Db/WebhookListener.php for an example, only special getters/setters are added for fields needing encryption/decryption, but for all others you should only add methods as phpdoc, Entity will manage.

come-nc commented on apps/cloud_federation_api/lib/Db/FederatedInviteMapper.php:
```diff
@@ -0,0 +1,96 @@
+use OCP\DB\QueryBuilder\IQueryBuilder;
+use OCP\IDBConnection;
+
+class FederatedInviteMapper extends QBMapper {
```
```suggestion
/**
 * @template-extends QBMapper<FederatedInvite>
 */
class FederatedInviteMapper extends QBMapper {
```

This will help with psalm

mickenordin replied:
done

mickenordin commented:
> I think because you defined getters/setters you short circtuited the entity/mapper system and had to re-implement stuff.
> 
> Look at https://github.com/nextcloud/server/blob/master/apps/webhook_listeners/lib/Db/WebhookListener.php for an example, only special getters/setters are added for fields needing encryption/decryption, but for all others you should only add methods as phpdoc, Entity will manage.

At least now I have something that works again, and I can accept handcrafted invites with curl, using the entity and mapper. :smile: 

I will look at your example and see if I can strip away excess and remove unnecessary cruft.

Once again, thank you so much for the review! :bow:

mickenordin commented on apps/cloud_federation_api/lib/Db/FederatedInvite.php:
```diff
@@ -0,0 +1,100 @@
+	}
+	public function getRecipientProvider(): ?string {
+		return $this->recipientProvider;
+	}
```
done

mickenordin commented:
I think that is all the comments addressed now.

mickenordin commented on apps/cloud_federation_api/lib/Controller/RequestHandlerController.php:
```diff
@@ -213,6 +225,103 @@ public function addShare($shareWith, $name, $description, $providerId, $owner, $
+			return new JSONResponse($response, $status);
+		}
+	    $unixstart = \DateTime::createFromFormat('U', '1');
+	    $expiredAt = \DateTime::createFromFormat('U', strval($invitation->getExpiredAt()));
```
I had to implement createFromFormat in ITimeFactory/TimeFactory as I could not find a useful way around it with the current implementation, so I have a commit for that too in here now.

mickenordin committed:
feat(OCM-invites): Implementation of invitation flow

This patchset:
* implements the /invite-accepted endpoint
* adds capabilities and inviteAceptDialog to the discovery
* adds a FederatedInviteAcceptedEvent

https://cs3org.github.io/OCM-API/docs.html?branch=v1.1.0&repo=OCM-API&user=cs3org#/paths/~1invite-accepted/post

Co-authored-by: Anna <anna@nextcloud.com>
Co-authored-by: Côme Chilliet <come.chilliet@nextcloud.com>
Co-authored-by: Joas Schilling <213943+nickvergessen@users.noreply.github.com>
Co-authored-by: Navid Shokri <navid.pdp11@gmail.com>
Signed-off-by: Micke Nordin <kano@sunet.se>

mickenordin commented on apps/cloud_federation_api/lib/Migration/Version1016Date202502262004.php:
```diff
@@ -0,0 +1,96 @@
+				'length' => 60,
+			]);
+			$table->addColumn('accepted', Types::BOOLEAN, [
+				'notnull' => false,
```
I reverted this, since at least sqlite backend refuses to store false if it is not allowed to store false. I left the bool as not nullable in the model and it works, so I dont think this is an issue.

mickenordin commented:
Thanks to @miaulalala too, very valuable. Should be addressed now.

come-nc commented on apps/cloud_federation_api/lib/Controller/RequestHandlerController.php:
```diff
@@ -213,6 +225,102 @@ public function addShare($shareWith, $name, $description, $providerId, $owner, $
+		$this->logger->debug('Invite accepted for ' . $userId . ' with token ' . $token . ' and email ' . $email . ' and name ' . $name);
+
+		$updated = $this->timeFactory->getTime();
+		/** @var IQueryBuilder $qb */
```
```suggestion
```

come-nc commented on apps/cloud_federation_api/lib/Controller/RequestHandlerController.php:
```diff
@@ -213,6 +225,102 @@ public function addShare($shareWith, $name, $description, $providerId, $owner, $
+			$invitation->setExpiredAt($updated);
+		}
+
+		if ($invitation->getExpiredAt() < $updated) {
```
```suggestion
		elseif ($invitation->getExpiredAt() < $updated) {
```

come-nc commented on apps/cloud_federation_api/lib/Controller/RequestHandlerController.php:
```diff
@@ -213,6 +225,102 @@ public function addShare($shareWith, $name, $description, $providerId, $owner, $
+		}
+		$unixstart = $this->timeFactory->createFromFormat('U', '1');
+		$expiredAt = $this->timeFactory->createFromFormat('U', strval($invitation->getExpiredAt()));
+		if ($expiredAt == $unixstart) {
+			$invitation->setExpiredAt($updated);
```
Can you add a comment explaining why this is done? It’s not clear to me at first read.
Why would expiredAt be equal to epoch + 1 (or epoch?), and why set it to $updated?

mickenordin replied:
The reason for doing this at all is that I want to support invites with no expiration date. Type juggling nightmares made me set invites without expiration to expire at 1 second after epoch. If you or someone else has ideas how to do this in a smarter/better way, it would be most welcome. I agree that the minimum should be a comment, but maybe we can come up with something better here. There is also the issue of strict equality check that @miaulalala noticed, and that I initially had, but had to roll back yesterday when it failed on me.

mickenordin replied:
I think it is worse than that. I can be null, and it can also be false, depending on db backend.

miaulalala replied:
you could check for `empty`: https://3v4l.org/6gH8F

That will allow you to both catch `null` and `false` values.

I think you're kinda stuck overthinking this :wink: 

If there's no expiration coming from the DB, you can simply proceed.

If there is a timestamp coming back, check against `$updated`.

So

```php

if (!empty($invitation->getExpiredAt()) && $updated > $invitation->getExpiredAt()) {
	// return an error here
}

// else proceed with the code as the invation is always valid
``

mickenordin replied:
Should be fixed now

come-nc commented on apps/cloud_federation_api/lib/Controller/RequestHandlerController.php:
```diff
@@ -213,6 +225,102 @@ public function addShare($shareWith, $name, $description, $providerId, $owner, $
+		$sharedFromEmail = $localUser->getPrimaryEMailAddress();
+		$sharedFromDisplayName = $localUser->getDisplayName();
+
+		$response = ['userID' => $invitation->getUserId(), 'email' => $sharedFromEmail, 'name' => $sharedFromDisplayName];
```
```suggestion
		$response = ['userID' => $localUser->getUID(), 'email' => $sharedFromEmail, 'name' => $sharedFromDisplayName];
```
This feels safer to me, as calling get on the user manager might get a user whose exact uid differs for some reason, depending on the backend used (like lowercase for instance).

come-nc commented on apps/cloud_federation_api/lib/Db/FederatedInviteMapper.php:
```diff
@@ -0,0 +1,37 @@
+		return $this->findEntity($qb);
+	}
+
+	public function dataArrayToInvite(array $data): FederatedInvite {
+		return $this->mapRowToEntity($data);
+	}
```
This is unused, no?

miaulalala commented on apps/cloud_federation_api/lib/Controller/RequestHandlerController.php:
```diff
@@ -213,6 +225,102 @@ public function addShare($shareWith, $name, $description, $providerId, $owner, $
+		// Note: Not implementing 404 Invitation token does not exist, instead using 400
+
+		if ($invitation->isAccepted() === true) {
+			$response = ['message' => 'Invite already accepted', 'error' => true];
+			$status = Http::STATUS_CONFLICT;
+			return new JSONResponse($response, $status);
```
**Nit**:

Does the spec necessitate a 409 response or could we simply treat this as a status 200?

nickvergessen replied:
I find this weird, maybe the connection got interrupted or something and the server simply retried later?

mickenordin replied:
This part is mandated by the spec, if the invitation has allready been accepted, we should return 409.

miaulalala commented on apps/cloud_federation_api/lib/Controller/RequestHandlerController.php:
```diff
@@ -213,6 +225,102 @@ public function addShare($shareWith, $name, $description, $providerId, $owner, $
+		}
+		$unixstart = $this->timeFactory->createFromFormat('U', '1');
+		$expiredAt = $this->timeFactory->createFromFormat('U', strval($invitation->getExpiredAt()));
+		if ($expiredAt == $unixstart) {
```
```suggestion
		if ($expiredAt === $unixstart) {
```
Please use strict comparison

mickenordin replied:
For some reason the check fails when doing that, see: 764d3e034b60af7cbad3492009e492382ad7f83e

miaulalala commented on apps/cloud_federation_api/lib/Db/FederatedInviteMapper.php:
```diff
@@ -0,0 +1,37 @@
+
+	public function __construct(IDBConnection $db) {
+		parent::__construct($db, self::TABLE_NAME);
+	}
+	public function findByToken(string $token): ?FederatedInvite {
```
```suggestion
	}
	
	public function findByToken(string $token): ?FederatedInvite {
```

miaulalala commented on apps/cloud_federation_api/lib/Migration/Version1016Date202502262004.php:
```diff
@@ -0,0 +1,93 @@
+		$schema = $schemaClosure();
+		$table_name = 'federated_invites';
+
+		if (! $schema->hasTable($table_name)) {
```
```suggestion
		if (!$schema->hasTable($table_name)) {
```

miaulalala reviewed (changes requested):
A couple small things

come-nc reviewed (commented):
Can you also run `composer run cs:fix` to fix code style?

mickenordin commented:
I have removed the clunky stuff around timestamps now, the result was much better I think.

BTW if you want to test this, you can create a share in your db:
```
insert into oc_federated_invites (user_id, token, created_at) values('admin', 'blahonga', 1742500700);
```
And then that invite can be accepted for example, like this:
```
curl -X POST "http://localhost:8000/index.php/ocm/invite-accepted?recipientProvider=https://sunet.drive.sunet.se&userId=kano@sunet.se&name=Micke+Nordin&token=blahonga&email=kano+spam@sunet.se" 
```
I have also run composer cs:fix again.

I realize that it would be good to add the above stuff as a test somewhere, but I am unfamiliar with the test suite, so pointers or help there is welcome, if you think a test should be included.

come-nc reviewed (commented):
All good from my side.

If you want to add a test for an API, I would say the best solution is a behat test in the build/integration folder.

miaulalala commented:
> I have removed the clunky stuff around timestamps now, the result was much better I think.
> 
> BTW if you want to test this, you can create a share in your db:
> 
> ```
> insert into oc_federated_invites (user_id, token, created_at) values('admin', 'blahonga', 1742500700);
> ```
> 
> And then that invite can be accepted for example, like this:
> 
> ```
> curl -X POST "http://localhost:8000/index.php/ocm/invite-accepted?recipientProvider=https://sunet.drive.sunet.se&userId=kano@sunet.se&name=Micke+Nordin&token=blahonga&email=kano+spam@sunet.se" 
> ```
> 
> I realize that it would be good to add the above stuff as a test somewhere, but I am unfamiliar with the test suite, so pointers or help there is welcome, if you think a test should be included. I suppose a test could also be added outside of this pr.
> 
> I also ran `composer run cs:fix` again.

If you would like to add tests, you can add a unit test to test the internal consistency for the `inviteAccepted` method. I would leave an integration test for now as it's complicated to set up (and you're not really doing that much except write to a table).

Let me know if you want some help, you can also join the [Community Chat](https://cloud.nextcloud.com/call/xs25tz5y) and ping me there if you have questions.

come-nc commented:
The openapi job is complaining about your controller:
```
Error: apps/cloud_federation_api/lib/Capabilities.php: No return value
Error: RequestHandler#inviteAccepted: @return: 'JSONResponse' needs 3 parameters
Error: RequestHandler#inviteAccepted: @param: recipientProvider: Missing description
Error: RequestHandler#inviteAccepted: @param: token: Missing description
Error: RequestHandler#inviteAccepted: @param: userId: Missing description
Error: RequestHandler#inviteAccepted: @param: email: Missing description
Error: RequestHandler#inviteAccepted: @param: name: Missing description
Error: RequestHandler#inviteAccepted: Returns no responses
PHP Fatal error:  Uncaught Error: app: Encountered 8 errors that need to be fixed!
```

mickenordin commented:
Crossing my fingers that openapi and psalm works now...

mickenordin commented:
> Crossing my fingers that openapi and psalm works now...

No such luck...

Seems the openapi script crashed in some horrible way?

mickenordin commented:
> > Crossing my fingers that openapi and psalm works now...
> 
> No such luck...
> 
> Seems the openapi script crashed in some horrible way?

Found the issue with the doc string in Capabilities

mickenordin commented:
I was also able to add a passing unit test, so from my side this is ready, wrt that.

come-nc reviewed (commented):
Small nitpick, otherwise all good.
The unit test does not use Mocks but I think that’s okay, it makes it kind of an integration test.

Would be better to use `\OCP\Server::get` instead of `OC::$server->get`.

come-nc commented:
Tests:
```
There were 2 failures:

1) OCA\CloudFederationApi\Tests\RequestHandlerControllerTest::testInviteAccepted
Failed asserting that two objects are equal.
--- Expected
+++ Actual
@@ @@
 OCP\AppFramework\Http\JSONResponse Object (
     'headers' => Array (...)
     'cookies' => Array ()
-    'status' => 200
+    'status' => 400
     'lastModified' => null
     'ETag' => null
     'contentSecurityPolicy' => null
     'featurePolicy' => null
-    'throttled' => false
+    'throttled' => true
     'throttleMetadata' => Array ()
     'data' => Array (
-        'userID' => 'admin'
-        'email' => null
-        'name' => 'admin'
+        'message' => 'Invalid or non existing token'
+        'error' => true
     )
     'encodeFlags' => 0
 )

/home/runner/actions-runner/_work/server/server/apps/cloud_federation_api/tests/RequestHandlerControllerTest.php:55

2) OCA\Federation\Tests\DbHandlerTest::testAddServer with data set #0 ('http://owncloud.org/', 'http://owncloud.org/', 'dc1e290686987885ec40126d85b19...55fb8b')
we need to start with a empty trusted_servers table
Failed asserting that an array is empty.

/home/runner/actions-runner/_work/server/server/apps/federation/tests/DbHandlerTest.php:50

mickenordin commented:
> Tests:
> 
> ```
> There were 2 failures:
> 
> 1) OCA\CloudFederationApi\Tests\RequestHandlerControllerTest::testInviteAccepted
> Failed asserting that two objects are equal.
> --- Expected
> +++ Actual
> @@ @@
>  OCP\AppFramework\Http\JSONResponse Object (
>      'headers' => Array (...)
>      'cookies' => Array ()
> -    'status' => 200
> +    'status' => 400
>      'lastModified' => null
>      'ETag' => null
>      'contentSecurityPolicy' => null
>      'featurePolicy' => null
> -    'throttled' => false
> +    'throttled' => true
>      'throttleMetadata' => Array ()
>      'data' => Array (
> -        'userID' => 'admin'
> -        'email' => null
> -        'name' => 'admin'
> +        'message' => 'Invalid or non existing token'
> +        'error' => true
>      )
>      'encodeFlags' => 0
>  )
> 
> /home/runner/actions-runner/_work/server/server/apps/cloud_federation_api/tests/RequestHandlerControllerTest.php:55
> 
> 2) OCA\Federation\Tests\DbHandlerTest::testAddServer with data set #0 ('http://owncloud.org/', 'http://owncloud.org/', 'dc1e290686987885ec40126d85b19...55fb8b')
> we need to start with a empty trusted_servers table
> Failed asserting that an array is empty.
> 
> /home/runner/actions-runner/_work/server/server/apps/federation/tests/DbHandlerTest.php:50
> ```

This is very strange, it passes locally:

```
NOCOVERAGE=1 ./autotest.sh sqlite apps/cloud_federation_api/tests/RequestHandlerControllerTest.php | tail -n 15 
Updated 0 paths from the index
Install background jobs
Nextcloud was successfully installed
Testing with sqlite ...
No coverage
/home/micke/sources/nextcloud/server/build/integration/vendor/bin/phpunit --colors=always --configuration phpunit-autotest.xml --log-junit autotest-results-sqlite.xml ../apps/cloud_federation_api/tests/RequestHandlerControllerTest.php 
PHPUnit 9.6.21 by Sebastian Bergmann and contributors.

Runtime:       PHP 8.4.4
Configuration: phpunit-autotest.xml

.                                                                   1 / 1 (100%)

Time: 00:00.009, Memory: 24.00 MB

OK (1 test, 1 assertion)
```

miaulalala commented:
> > Tests:
> > ```
> > There were 2 failures:
> > 
> > 1) OCA\CloudFederationApi\Tests\RequestHandlerControllerTest::testInviteAccepted
> > Failed asserting that two objects are equal.
> > --- Expected
> > +++ Actual
> > @@ @@
> >  OCP\AppFramework\Http\JSONResponse Object (
> >      'headers' => Array (...)
> >      'cookies' => Array ()
> > -    'status' => 200
> > +    'status' => 400
> >      'lastModified' => null
> >      'ETag' => null
> >      'contentSecurityPolicy' => null
> >      'featurePolicy' => null
> > -    'throttled' => false
> > +    'throttled' => true
> >      'throttleMetadata' => Array ()
> >      'data' => Array (
> > -        'userID' => 'admin'
> > -        'email' => null
> > -        'name' => 'admin'
> > +        'message' => 'Invalid or non existing token'
> > +        'error' => true
> >      )
> >      'encodeFlags' => 0
> >  )
> > 
> > /home/runner/actions-runner/_work/server/server/apps/cloud_federation_api/tests/RequestHandlerControllerTest.php:55
> > 
> > 2) OCA\Federation\Tests\DbHandlerTest::testAddServer with data set #0 ('http://owncloud.org/', 'http://owncloud.org/', 'dc1e290686987885ec40126d85b19...55fb8b')
> > we need to start with a empty trusted_servers table
> > Failed asserting that an array is empty.
> > 
> > /home/runner/actions-runner/_work/server/server/apps/federation/tests/DbHandlerTest.php:50
> > ```
> 
> This is very strange, it passes locally:
> 
> ```
> NOCOVERAGE=1 ./autotest.sh sqlite apps/cloud_federation_api/tests/RequestHandlerControllerTest.php | tail -n 15 
> Updated 0 paths from the index
> Install background jobs
> Nextcloud was successfully installed
> Testing with sqlite ...
> No coverage
> /home/micke/sources/nextcloud/server/build/integration/vendor/bin/phpunit --colors=always --configuration phpunit-autotest.xml --log-junit autotest-results-sqlite.xml ../apps/cloud_federation_api/tests/RequestHandlerControllerTest.php 
> PHPUnit 9.6.21 by Sebastian Bergmann and contributors.
> 
> Runtime:       PHP 8.4.4
> Configuration: phpunit-autotest.xml
> 
> .                                                                   1 / 1 (100%)
> 
> Time: 00:00.009, Memory: 24.00 MB
> 
> OK (1 test, 1 assertion)
> ```

It's probably bc you run a single test locally but the full suite is runnig in CI

mickenordin commented:
@nickvergessen would you care to review again, would be nice if we could get this merged before there crops up any merge conflicts :)

nickvergessen commented:
- I triggered a CI run with this change in Talk, so Talks federation is confirmed to continue working:
https://github.com/nextcloud/spreed/pull/14765

nickvergessen commented on apps/cloud_federation_api/lib/Controller/RequestHandlerController.php:
```diff
@@ -213,6 +226,99 @@ public function addShare($shareWith, $name, $description, $providerId, $owner, $
+			$response->throttle();
+			return $response;
+		}
+		if (!$this->trustedServers->isTrustedServer($recipientProvider)) {
```
Is this correct? I don't see/understand why a server doing OCM would require to be a trusted Nextcloud Server? Trusting a server has a huge impact (e.g. sharing all user ids and names, etc), so having this as a prerequirement sounds weird?

mickenordin replied:
It is indeed optional in the spec, do you prefer that it be removed completely, or put behind a config option?

mickenordin replied:
An invitation is not the same as a share, but I get your meaning, it should be consistent.

mickenordin replied:
I removed the trusted server part now

nickvergessen commented on apps/cloud_federation_api/lib/Controller/RequestHandlerController.php:
```diff
@@ -213,6 +226,99 @@ public function addShare($shareWith, $name, $description, $providerId, $owner, $
+	#[PublicPage]
+	#[NoCSRFRequired]
+	#[BruteForceProtection(action: 'inviteAccepted')]
+	public function inviteAccepted(string $recipientProvider, string $token, string $userId, string $email, string $name): JSONResponse {
```
So this is for Nextcloud to receive the "invite accepted" from other OCM share providers. But I see no request being added that makes us also send this request?

mickenordin replied:
This needs to be done in a web gui, the idea is to implement this in the contacts app. That part is not part of the OCM specification and is implementation specific.

mickenordin commented:
> I find this weird, maybe the connection got interrupted or something and the server simply retried later?
This is nevertheless how I read the specification, you should not accept an invite twice.

@nickvergessen I think I have answered the questions you had, is there anything more before we can get this merged? There are other pr's waiting for this to be merged.

mickenordin commented:
@ArtificialOwl I heard a rumor that you had volunteered to review :). Would you mind?

ArtificialOwl commented on lib/composer/composer/autoload_static.php:
```diff
@@ -21,6 +21,10 @@ class ComposerStaticInit749170dad3f5e7f9ca158f5a9f04f6a2
+        'B' => 
+        array (
+            'Bamarni\\Composer\\Bin\\' => 21,
+        ),
```
not sure this is needed

come-nc replied:
This should be fixed by running `composer install --no-dev` if I remember correctly

mickenordin replied:
Fixed

ArtificialOwl commented on lib/composer/composer/autoload_psr4.php:
```diff
@@ -10,5 +10,6 @@
     'OC\\' => array($baseDir . '/lib/private'),
     'OCP\\' => array($baseDir . '/lib/public'),
     'NCU\\' => array($baseDir . '/lib/unstable'),
+    'Bamarni\\Composer\\Bin\\' => array($vendorDir . '/bamarni/composer-bin-plugin/src'),
```
not sure this is needed

mickenordin replied:
Fixed

ArtificialOwl commented on lib/composer/composer/autoload_static.php:
```diff
@@ -40,6 +44,10 @@ class ComposerStaticInit749170dad3f5e7f9ca158f5a9f04f6a2
+        'Bamarni\\Composer\\Bin\\' => 
+        array (
+            0 => __DIR__ . '/..' . '/bamarni/composer-bin-plugin/src',
+        ),
```
same, I have not found its usefulness

mickenordin replied:
Fixed

ArtificialOwl reviewed (changes requested):
requesting only minor changes; tried it and seems to be working, based on my limited knowledge on the feature.

ArtificialOwl commented on apps/cloud_federation_api/lib/Db/FederatedInvite.php:
```diff
@@ -0,0 +1,92 @@
+	/**
+	 * @var bool $accepted
+	 */
+	protected $accepted;
```
You should be able to type the variables (without facing too much limitation) while keeping the `Entity` working.
Have a look into `lib/unstable/Security/Signature/Model/Signatory.php` for a recent example

mickenordin commented:
@ArtificialOwl your comments should be addressed now I think.

mickenordin commented on lib/private/OCM/Model/OCMProvider.php:
```diff
@@ -88,6 +118,34 @@ public function getEndPoint(): string {
+	 *
+	 * @return $this
+	 */
+	public function setCapabilities(array $capabilities): static {
+		foreach ($capabilities as $value) {
+			if (!in_array($value, $this->capabilities)) {
+				array_push($this->capabilities, $value);
+			}
+		}
+
+		return $this;
+	}
+
+	/**
+	 * @return array
+	 */
+	public function getCapabilities(): array {
+		return $this->capabilities;
+	}
```
But I see what you mean, adding it to the interface will break compatibility for the implementors. So yeah, a new interface might be the way to go.

ArtificialOwl commented:
Hello Micke, 
Thanks again for your patience,

I think we're good on this PR, still have a last question regarding `ICapabilityAwareOCMProvider`; has it been requested to create a new interface ? adding the method to the previous `IOCMProvider` is not doable ?

miaulalala reviewed (commented):
@ArtificialOwl we went through the code in a call and I approve but I leave final approval to you :)

ArtificialOwl commented:
> > Hello Micke, Thanks again for your patience,
> > I think we're good on this PR, still have a last question regarding `ICapabilityAwareOCMProvider`; has it been requested to create a new interface ? adding the method to the previous `IOCMProvider` is not doable ?
> 
> It would break backwards compatibility, so extending the `IOCMProvider` seemed like the best choice. Do you disagree?

I dont think it would break anything, but I might be wrong :)

mickenordin commented:
@ArtificialOwl @miaulalala who can merge the PR now, if all the requirements are met?

AndyScherzinger commented:
@mickenordin @ArtificialOwl @sorbaugh 

accordign to @SystemKeeper this breaks talk federation at the moment, as `IOCMProvider` was replaced by `ICapabilityAwareOCMProvider`. Can someone tell me if that is intended to break? I see `IOCMProvider` used in talk, circles and groupfolders and it is currently not documented at https://docs.nextcloud.com/server/latest/developer_manual/app_publishing_maintenance/app_upgrade_guide/upgrade_to_32.html

"replaced".. at least the alias was removed: https://github.com/nextcloud/server/pull/51113/files#diff-b868f3fc169b78b9222fb4822eba2940dc6eccb4b8ea9df464366be45f059ef8

Accordign to @st3iny the old alias should stay. Should be fine since the new interface extends the old one: `interface ICapabilityAwareOCMProvider extends IOCMProvider {`

Perhaps add a deprecation notice to `IOCMProvider`?

Also cc @icewind1991 and @nickvergessen given the impacted apps.

mickenordin commented:
There was not an intention for it to break. So the proposed solution is to revert this change?

```

-		$this->registerAlias(IOCMProvider::class, OCMProvider::class);
+		$this->registerAlias(ICapabilityAwareOCMProvider::class, OCMProvider::class);
```

mickenordin commented:
Relevant comments are here:

https://github.com/nextcloud/server/pull/51113#issuecomment-2930175857
https://github.com/nextcloud/server/pull/51113#issuecomment-2930222823
https://github.com/nextcloud/server/pull/51113#issuecomment-2938985293

st3iny commented:
> Can they have the same name? I.e, can we do:
> 
> ```
> 		$this->registerAlias(IOCMProvider::class, OCMProvider::class);
> 		$this->registerAlias(ICapabilityAwareOCMProvider::class, OCMProvider::class);
> ```

Yes, that should work.

I have another question: Is there a reason to split the logic in to two interfaces and keep both? Another possibility to just add the new methods from `ICapabilityAwareOCMProvider` to `IOCMProvider`. Do you see a reason why this would be a bad idea? (Asking because I lack a bit of context about the PR.)

mickenordin commented:
If we can quickly decide on the way we want to move forward, I can draft a PR for a fix.

nickvergessen commented on lib/private/Server.php:
```diff
@@ -1271,7 +1272,7 @@ public function __construct($webRoot, \OC\Config $config) {
 
 		$this->registerAlias(IPhoneNumberUtil::class, PhoneNumberUtil::class);
 
-		$this->registerAlias(IOCMProvider::class, OCMProvider::class);
```
This is a breaking change. The old and new alias need to be in place

AndyScherzinger replied:
https://github.com/nextcloud/server/pull/53465
